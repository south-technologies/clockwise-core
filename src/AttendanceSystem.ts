interface Session {
  startTime: Date;
  endTime: Date;
  sessionHours: number;
}

interface FormattedSession {
  startTime: Date;
  endTime: Date;
  sessionHours: string;
}

interface FormattedAttendanceData {
  [date: string]: {
      morning: FormattedSession | null;
      afternoon: FormattedSession | null;
      dailyHours: string | null;
  };
}

export class AttendanceSystem {
  attendanceData: { [date: string]: { morning: Session | null; afternoon: Session | null; dailyHours: number | null } };
  currentDay: string;
  currentSession: string | null;
  creditBalance: number;
  debitBalance: number;
  theoreticalWeeklyAttendanceTime: number;
  dailyHours: number;

  constructor() {
      this.attendanceData = {};
      this.currentDay = "";
      this.currentSession = null;
      this.creditBalance = 0;
      this.debitBalance = 0;
      this.theoreticalWeeklyAttendanceTime = 0;
      this.dailyHours = 0;
  }

  // Format time in HH:mm format
  formatTime(hours: number): string {
      const hoursFormatted = Math.floor(hours);
      const minutesFormatted = Math.floor((hours - hoursFormatted) * 60);
      return `${hoursFormatted.toString().padStart(2, '0')}:${minutesFormatted.toString().padStart(2, '0')}`;
  }

  // Start a new day
  startDay(date: Date): void {
      const formattedDate = date.toDateString();
      this.attendanceData[formattedDate] = { morning: null, afternoon: null, dailyHours: null };
      this.currentDay = formattedDate;
  }

  // Start a new session (morning or afternoon)
  startSession(session: string): void {
      this.currentSession = session;
  }

  // End the current session and calculate session hours
  endSession(startTime: Date, endTime: Date): void {
      if (!this.attendanceData[this.currentDay]) {
          this.attendanceData[this.currentDay] = { morning: null, afternoon: null, dailyHours: 0 };
      }
      const sessionHours = (endTime.getTime() - startTime.getTime()) / (1000 * 60 * 60);
      const currentDayData = this.attendanceData[this.currentDay] as { morning: Session | null; afternoon: Session | null; dailyHours: number };
      currentDayData[this.currentSession!] = { startTime, endTime, sessionHours };
      currentDayData.dailyHours += sessionHours;
  }

  // Calculate the total hours for a given day
  calculateDayHours(day: string): void {
      const sessions = this.attendanceData[day];
      const morningDiff = sessions.morning!.endTime.getTime() - sessions.morning!.startTime.getTime();
      const afternoonDiff = sessions.afternoon!.endTime.getTime() - sessions.afternoon!.startTime.getTime();
      const totalHours = (morningDiff + afternoonDiff) / (1000 * 60 * 60);

      if (totalHours >= 8) {
          this.creditBalance += totalHours - 8;
      } else {
          this.debitBalance += 8 - totalHours;
      }
  }

  // Adjust theoretical weekly attendance time based on credit and debit balances
  adjustTheoreticalHours(day: string): void {
      this.theoreticalWeeklyAttendanceTime += 8;
      const adjustment = this.creditBalance - 8;
      if (adjustment > 0) {
          this.theoreticalWeeklyAttendanceTime -= adjustment;
          this.creditBalance -= adjustment;
      }
  }

  // Convert the attendance data to a formatted JSON object
  toJSON(): {
      attendanceData: FormattedAttendanceData;
      creditBalance: string;
      debitBalance: string;
      CDBalance: string;
      theoreticalWeeklyAttendanceTime: number;
  } {
      const formattedAttendanceData: FormattedAttendanceData = {};

      for (const date in this.attendanceData) {
          const sessions = this.attendanceData[date];
          const formattedSessions: { morning: FormattedSession | null; afternoon: FormattedSession | null; dailyHours: string | null } = {
              morning: null,
              afternoon: null,
              dailyHours: null,
          };

          for (const sessionType in sessions) {
              const session = sessions[sessionType];
              if (session && session.sessionHours) {
                  formattedSessions[sessionType] = {
                      startTime: session.startTime,
                      endTime: session.endTime,
                      sessionHours: this.formatTime(session.sessionHours),
                  };
              }
          }

          formattedAttendanceData[date] = {
              ...formattedSessions,
              dailyHours: this.formatTime(sessions.dailyHours! || 0), // Provide default value of 0 if dailyHours is null
          };
      }

      const creditHours = Math.floor(this.creditBalance);
      const creditMinutes = Math.floor((this.creditBalance - creditHours) * 60);

      const debitHours = Math.floor(this.debitBalance);
      const debitMinutes = Math.floor((this.debitBalance - debitHours) * 60);

      let CDBalanceHours = Math.floor(this.debitBalance - this.creditBalance);
      let CDBalanceMinutes = Math.abs(Math.floor((this.debitBalance - this.creditBalance) * 60));

      if (CDBalanceHours < 0 || (CDBalanceHours === 0 && this.creditBalance < this.debitBalance)) {
          CDBalanceHours *= -1;
          CDBalanceMinutes *= -1;
      }

      return {
          attendanceData: formattedAttendanceData,
          creditBalance: `${creditHours.toString().padStart(2, '0')}:${creditMinutes.toString().padStart(2, '0')}`,
          debitBalance: `-${debitHours.toString().padStart(2, '0')}:${debitMinutes.toString().padStart(2, '0')}`,
          CDBalance: `${CDBalanceHours.toString().padStart(2, '0')}:${CDBalanceMinutes.toString().padStart(2, '0')}`,
          theoreticalWeeklyAttendanceTime: this.theoreticalWeeklyAttendanceTime,
      };
  }
}
