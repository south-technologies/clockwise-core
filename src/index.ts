export { AttendanceSystem } from './AttendanceSystem.ts';
