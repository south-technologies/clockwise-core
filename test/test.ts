import { AttendanceSystem } from '../src/AttendanceSystem';

describe('AttendanceSystem', () => {
  let attendanceSystem: AttendanceSystem;

  beforeEach(() => {
    attendanceSystem = new AttendanceSystem();
  });

  test('should process days correctly', () => {
    // Day 1
    const startDate1 = new Date('2024-03-05');
    attendanceSystem.startDay(startDate1);

    attendanceSystem.startSession('morning');
    attendanceSystem.endSession(new Date('2024-03-05T08:30:00'), new Date('2024-03-05T12:30:00'));

    attendanceSystem.startSession('afternoon');
    attendanceSystem.endSession(new Date('2024-03-05T13:30:00'), new Date('2024-03-05T17:30:00'));

    attendanceSystem.calculateDayHours(startDate1.toDateString());
    attendanceSystem.adjustTheoreticalHours(startDate1.toDateString());

    // Day 2
    const startDate2 = new Date('2024-03-06');
    attendanceSystem.startDay(startDate2);

    attendanceSystem.startSession('morning');
    attendanceSystem.endSession(new Date('2024-03-06T08:00:00'), new Date('2024-03-06T12:00:00'));

    attendanceSystem.startSession('afternoon');
    attendanceSystem.endSession(new Date('2024-03-06T13:00:00'), new Date('2024-03-06T17:00:00'));

    attendanceSystem.calculateDayHours(startDate2.toDateString());
    attendanceSystem.adjustTheoreticalHours(startDate2.toDateString());

    // Day 3
    const startDate3 = new Date('2024-03-07');
    attendanceSystem.startDay(startDate3);

    attendanceSystem.startSession('morning');
    attendanceSystem.endSession(new Date('2024-03-07T08:00:00'), new Date('2024-03-07T12:00:00'));

    attendanceSystem.startSession('afternoon');
    attendanceSystem.endSession(new Date('2024-03-07T13:10:00'), new Date('2024-03-07T17:00:00'));

    attendanceSystem.calculateDayHours(startDate3.toDateString());
    attendanceSystem.adjustTheoreticalHours(startDate3.toDateString());

    expect(attendanceSystem.toJSON()).toMatchSnapshot();
  });
});
