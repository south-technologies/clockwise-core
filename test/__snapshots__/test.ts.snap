// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AttendanceSystem should process days correctly 1`] = `
{
  "CDBalance": "00:-10",
  "attendanceData": {
    "Thu Mar 07 2024": {
      "afternoon": {
        "endTime": 2024-03-07T16:00:00.000Z,
        "sessionHours": "03:50",
        "startTime": 2024-03-07T12:10:00.000Z,
      },
      "dailyHours": "07:50",
      "morning": {
        "endTime": 2024-03-07T11:00:00.000Z,
        "sessionHours": "04:00",
        "startTime": 2024-03-07T07:00:00.000Z,
      },
    },
    "Tue Mar 05 2024": {
      "afternoon": {
        "endTime": 2024-03-05T16:30:00.000Z,
        "sessionHours": "04:00",
        "startTime": 2024-03-05T12:30:00.000Z,
      },
      "dailyHours": "08:00",
      "morning": {
        "endTime": 2024-03-05T11:30:00.000Z,
        "sessionHours": "04:00",
        "startTime": 2024-03-05T07:30:00.000Z,
      },
    },
    "Wed Mar 06 2024": {
      "afternoon": {
        "endTime": 2024-03-06T16:00:00.000Z,
        "sessionHours": "04:00",
        "startTime": 2024-03-06T12:00:00.000Z,
      },
      "dailyHours": "08:00",
      "morning": {
        "endTime": 2024-03-06T11:00:00.000Z,
        "sessionHours": "04:00",
        "startTime": 2024-03-06T07:00:00.000Z,
      },
    },
  },
  "creditBalance": "00:00",
  "debitBalance": "-00:10",
  "theoreticalWeeklyAttendanceTime": 24,
}
`;
